// @consta/uikit/__internal__/cjs-src/components/User/User.js
"use strict";
exports.__esModule = true;
exports.User = exports.cnUser = exports.userPropStatus = exports.userPropWidthDefault = exports.userPropWidth = exports.userPropViewDefault = exports.userPropView = exports.userPropSizeDefault = exports.userPropSize = void 0;
var tslib_1 = require("tslib");
require("./User.css");
var react_1 = tslib_1.__importDefault(require("react"));
var IconSelect_1 = require("../../icons/IconSelect/IconSelect");
var bem_1 = require("../../utils/bem");
var getSizeByMap_1 = require("../../utils/getSizeByMap");
var PropsWithAsAttributes_1 = require("../../utils/types/PropsWithAsAttributes");
var Avatar_1 = require("../Avatar/Avatar");
var Text_1 = require("../Text/Text");
exports.userPropSize = ['m', 's'];
exports.userPropSizeDefault = exports.userPropSize[0];
exports.userPropView = ['clear', 'ghost'];
exports.userPropViewDefault = exports.userPropView[0];
exports.userPropWidth = ['default', 'full'];
exports.userPropWidthDefault = exports.userPropWidth[0];
exports.userPropStatus = ['available', 'remote', 'out'];
exports.cnUser = bem_1.cn('User');
var infoSizeMap = {
    s: '2xs',
    m: 'xs'
};
var arrowSizeMap = {
    s: 'xs',
    m: 's'
};
exports.User = PropsWithAsAttributes_1.forwardRefWithAs(function (props, ref) {
    var _a = props.as, as = _a === void 0 ? 'div' : _a, className = props.className, _b = props.size, size = _b === void 0 ? exports.userPropSizeDefault : _b, avatarUrl = props.avatarUrl, name = props.name, _c = props.view, view = _c === void 0 ? exports.userPropViewDefault : _c, _d = props.width, width = _d === void 0 ? exports.userPropWidthDefault : _d, propOnlyAvatar = props.onlyAvatar, withArrow = props.withArrow, info = props.info, status = props.status, otherProps = tslib_1.__rest(props, ["as", "className", "size", "avatarUrl", "name", "view", "width", "onlyAvatar", "withArrow", "info", "status"]);
    var Tag = as;
    var onlyAvatar = propOnlyAvatar || (!name && !info);
    var infoSize = getSizeByMap_1.getSizeByMap(infoSizeMap, size);
    var arrowSize = getSizeByMap_1.getSizeByMap(arrowSizeMap, size);
    return (react_1["default"].createElement(Tag, tslib_1.__assign({}, otherProps, { className: exports.cnUser({ size: size, view: view, width: width, withArrow: withArrow, minified: onlyAvatar }, [className]), ref: ref }),
        react_1["default"].createElement("div", { className: exports.cnUser('AvatarWrapper', { status: status }) },
            react_1["default"].createElement(Avatar_1.Avatar, { className: exports.cnUser('Avatar', { status: status }), size: size, url: avatarUrl, name: name })),
        !onlyAvatar && (name || info) && (react_1["default"].createElement("div", { className: exports.cnUser('Block') },
            name && (react_1["default"].createElement(Text_1.Text, { className: exports.cnUser('Name'), size: size, view: "primary", lineHeight: "2xs" }, name)),
            info && (react_1["default"].createElement(Text_1.Text, { className: exports.cnUser('Info'), size: infoSize, view: "secondary", lineHeight: "2xs" }, info)))),
        withArrow && react_1["default"].createElement(IconSelect_1.IconSelect, { className: exports.cnUser('Arrow'), size: arrowSize, view: "secondary" })));
});
