// @consta/uikit/__internal__/cjs-src/components/MultiCombobox/MultiCombobox.js
"use strict";
exports.__esModule = true;
exports.MultiCombobox = exports.multiComboboxPropSizeDefault = exports.multiComboboxPropSize = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var useSelect_1 = require("../../hooks/useSelect/useSelect");
var IconClose_1 = require("../../icons/IconClose/IconClose");
var IconSelect_1 = require("../../icons/IconSelect/IconSelect");
var MixFocus_1 = require("../../mixs/MixFocus/MixFocus");
var scrollIntoView_1 = require("../../utils/scrollIntoView");
var cnSelect_1 = require("../SelectComponents/cnSelect");
var helpers_1 = require("../SelectComponents/helpers");
var SelectContainer_1 = require("../SelectComponents/SelectContainer/SelectContainer");
var SelectDropdown_1 = require("../SelectComponents/SelectDropdown/SelectDropdown");
var SelectValueTag_1 = require("../SelectComponents/SelectValueTag/SelectValueTag");
var types_1 = require("../SelectComponents/types");
exports.multiComboboxPropSize = ['m', 's', 'l'];
exports.multiComboboxPropSizeDefault = exports.multiComboboxPropSize[0];
exports.MultiCombobox = function (props) {
    var defaultOptionsRef = react_1.useRef(null);
    var placeholder = props.placeholder, onBlur = props.onBlur, onFocus = props.onFocus, options = props.options, onChange = props.onChange, value = props.value, getOptionLabel = props.getOptionLabel, disabled = props.disabled, ariaLabel = props.ariaLabel, id = props.id, _a = props.form, form = _a === void 0 ? types_1.DefaultPropForm : _a, _b = props.view, view = _b === void 0 ? types_1.DefaultPropView : _b, _c = props.size, size = _c === void 0 ? exports.multiComboboxPropSizeDefault : _c, onCreate = props.onCreate, getGroupOptions = props.getGroupOptions, _d = props.labelForCreate, labelForCreate = _d === void 0 ? 'Добавить' : _d, _e = props.labelForNotFound, labelForNotFound = _e === void 0 ? 'Не найдено' : _e, dropdownClassName = props.dropdownClassName, _f = props.dropdownRef, dropdownRef = _f === void 0 ? defaultOptionsRef : _f, name = props.name, restProps = tslib_1.__rest(props, ["placeholder", "onBlur", "onFocus", "options", "onChange", "value", "getOptionLabel", "disabled", "ariaLabel", "id", "form", "view", "size", "onCreate", "getGroupOptions", "labelForCreate", "labelForNotFound", "dropdownClassName", "dropdownRef", "name"]);
    var _g = react_1.useState(false), isFocused = _g[0], setIsFocused = _g[1];
    var _h = react_1.useState({
        value: ''
    }), inputData = _h[0], setInputData = _h[1];
    var toggleRef = react_1.useRef(null);
    var controlInnerRef = react_1.useRef(null);
    var helperInputFakeElement = react_1.useRef(null);
    var handlerChangeValue = function (v) {
        var _a;
        if (typeof onChange === 'function' && v) {
            onChange(v);
        }
        setInputData({ value: (_a = toggleRef.current) === null || _a === void 0 ? void 0 : _a.value });
    };
    var controlRef = react_1.useRef(null);
    var arrValue = typeof value !== 'undefined' && value !== null ? tslib_1.__spreadArrays(value) : null;
    var hasGroup = typeof getGroupOptions === 'function';
    var scrollToIndex = function (index) {
        if (!dropdownRef.current) {
            return;
        }
        var elements = dropdownRef.current.querySelectorAll('div[role=option]');
        scrollIntoView_1.scrollIntoView(elements[index], dropdownRef.current);
    };
    var onSelectOption = function () {
        setInputData({ value: '' });
    };
    var _j = useSelect_1.useSelect({
        options: options,
        value: arrValue,
        onChange: handlerChangeValue,
        optionsRef: dropdownRef,
        controlRef: controlRef,
        scrollToIndex: scrollToIndex,
        disabled: disabled,
        getOptionLabel: getOptionLabel,
        onCreate: onCreate,
        getGroupOptions: getGroupOptions,
        multi: true,
        onSelectOption: onSelectOption
    }), visibleOptions = _j.visibleOptions, highlightedIndex = _j.highlightedIndex, getToggleProps = _j.getToggleProps, getOptionProps = _j.getOptionProps, isOpen = _j.isOpen, setOpen = _j.setOpen;
    var handleInputFocus = function (e) {
        if (!disabled) {
            if (!isFocused) {
                setIsFocused(true);
            }
            if (typeof onFocus === 'function') {
                onFocus(e);
            }
        }
    };
    var handleInputBlur = function (e) {
        var _a;
        if (isOpen) {
            (_a = toggleRef.current) === null || _a === void 0 ? void 0 : _a.focus();
            return;
        }
        if (isFocused) {
            setIsFocused(false);
            if (typeof onBlur === 'function') {
                onBlur(e);
            }
        }
    };
    var handleToggleDropdown = function () {
        var _a;
        if (isOpen) {
            setOpen(false);
            setIsFocused(false);
        }
        else {
            setOpen(true);
            setIsFocused(true);
            (_a = toggleRef.current) === null || _a === void 0 ? void 0 : _a.focus();
        }
    };
    var handleClear = function () {
        setInputData({ value: '' });
        typeof onChange === 'function' && onChange(null);
        setIsFocused(false);
    };
    var handleClearButtonFocus = function () {
        setIsFocused(true);
    };
    var handleClearButtonBlur = function () {
        setIsFocused(false);
    };
    var handleInputChange = function () {
        var _a, _b;
        if (!isOpen) {
            setOpen(true);
        }
        var inputValue = (_b = (_a = toggleRef.current) === null || _a === void 0 ? void 0 : _a.value) !== null && _b !== void 0 ? _b : '';
        setInputData({ value: inputValue });
    };
    var handleControlClick = function () {
        var _a;
        (_a = toggleRef.current) === null || _a === void 0 ? void 0 : _a.focus();
        if (!isFocused) {
            setOpen(!isOpen);
        }
    };
    var showPlaceholder = (!(arrValue === null || arrValue === void 0 ? void 0 : arrValue.length) && inputData.value === '') || (arrValue === null && inputData.value === '');
    var handleCreate = function () {
        var _a;
        if (typeof onCreate === 'function') {
            var newValue = (_a = toggleRef.current) === null || _a === void 0 ? void 0 : _a.value.trim();
            newValue && onCreate(newValue);
        }
    };
    var handleRemoveValue = function (option) {
        var newVal = arrValue === null || arrValue === void 0 ? void 0 : arrValue.filter(function (arrVal) { return getOptionLabel(arrVal) !== getOptionLabel(option); });
        handlerChangeValue(newVal);
    };
    var getInputStyle = function () {
        if (!controlInnerRef.current || !helperInputFakeElement.current) {
            return {
                width: 0
            };
        }
        var fakeElWidth = helperInputFakeElement.current.offsetWidth + 20;
        var maxWidth = controlInnerRef.current ? controlInnerRef.current.offsetWidth - 15 : 0;
        var width = fakeElWidth > maxWidth ? maxWidth : fakeElWidth;
        return {
            width: width
        };
    };
    var inputStyle = react_1["default"].useMemo(function () { return getInputStyle(); }, [inputData.value, arrValue]);
    return (react_1["default"].createElement(SelectContainer_1.SelectContainer, tslib_1.__assign({ focused: isFocused, disabled: disabled, size: size, view: view, form: form, multi: true }, restProps),
        react_1["default"].createElement("div", { className: cnSelect_1.cnSelect('Control', { hasInput: true }), ref: controlRef, "aria-expanded": isOpen, "aria-haspopup": "listbox", id: id },
            react_1["default"].createElement("div", { className: cnSelect_1.cnSelect('ControlInner'), onClick: handleControlClick, onKeyDown: handleControlClick, role: "button", ref: controlInnerRef },
                react_1["default"].createElement("div", { className: cnSelect_1.cnSelect('ControlValueContainer') },
                    react_1["default"].createElement("div", { className: cnSelect_1.cnSelect('ControlValue') }, arrValue === null || arrValue === void 0 ? void 0 :
                        arrValue.map(function (option) {
                            var label = getOptionLabel(option);
                            var handleRemove = function (e) {
                                e.stopPropagation();
                                handleRemoveValue(option);
                            };
                            return (react_1["default"].createElement(SelectValueTag_1.SelectValueTag, { label: label, key: label, size: size, disabled: Boolean(disabled), handleRemove: handleRemove }));
                        }),
                        react_1["default"].createElement("input", tslib_1.__assign({}, getToggleProps({ onChange: handleInputChange }), { type: "text", name: name, onFocus: handleInputFocus, onBlur: handleInputBlur, "aria-label": ariaLabel, ref: toggleRef, value: inputData.value, className: cnSelect_1.cnSelect('Input', { size: size }), style: inputStyle })),
                        react_1["default"].createElement("div", { className: cnSelect_1.cnSelect('Placeholder', { isHidden: !showPlaceholder }) }, placeholder)))),
            react_1["default"].createElement("div", { className: cnSelect_1.cnSelect('Indicators') },
                arrValue && arrValue.length !== 0 && (react_1["default"].createElement("button", { type: "button", onClick: handleClear, className: cnSelect_1.cnSelect('ClearIndicator', [MixFocus_1.cnMixFocus()]), onFocus: handleClearButtonFocus, onBlur: handleClearButtonBlur },
                    react_1["default"].createElement(IconClose_1.IconClose, { size: "xs", className: cnSelect_1.cnSelect('ClearIndicatorIcon') }))),
                react_1["default"].createElement("div", { className: cnSelect_1.cnSelect('Delimiter') }),
                react_1["default"].createElement("button", { type: "button", className: cnSelect_1.cnSelect('IndicatorsDropdown'), tabIndex: -1, onClick: handleToggleDropdown },
                    react_1["default"].createElement(IconSelect_1.IconSelect, { size: "xs", className: cnSelect_1.cnSelect('DropdownIndicatorIcon') })))),
        react_1["default"].createElement(SelectDropdown_1.SelectDropdown, { isOpen: isOpen, size: size, controlRef: controlRef, visibleOptions: visibleOptions, highlightedIndex: highlightedIndex, getOptionProps: getOptionProps, onCreate: handleCreate, dropdownRef: dropdownRef, inputValue: inputData.value, id: id, hasGroup: hasGroup, selectedValues: arrValue, labelForCreate: labelForCreate, labelForNotFound: labelForNotFound, multi: true, getOptionLabel: getOptionLabel, form: helpers_1.getSelectDropdownForm(form), className: dropdownClassName }),
        react_1["default"].createElement("div", { className: cnSelect_1.cnSelect('HelperInputFakeElement'), ref: helperInputFakeElement }, inputData.value)));
};
